### ISSUE TITLE
Emerald Transport Feature Requests

### ISSUE THREAD

**Issue by LuigiHutch** [BuildCraft/BuildCraft#3033]
BC 7.0.25

Bug:
 - Changing the 'mode' of an emerald pipe when the pipe is being powered (by an autarchic gate/engine), it does not result in the pipe changing 'modes'. This means the pipe has to be unpowered for the pipe to change state. I would conside this a bug :) *(Fixed in 7.1.6. Please use separate issues for bugs and suggestions next time ~asie)*

Feature Requests:
 - Add a partial round robin mode: Currently the emerald pipe on round robin mode will only extract if ALL items set in emeralds pipe filter are in the adjacent inventory, a partial rond robin mode would extract in a round robin fashion if ANY of the items that are in the filter are in the adjacent chests

- Add a Blacklist round robin mode: Currently it only works in a whitelist fashion and a blacklist version could be quite useful.

**Comment by SandGrainOne** [comment]
These things sounds like ideas from somone who has been testing the emerald pipe without any real goals and not a player in real need of a feature. ;)

Please provide functional context and we will consider the options.

**Comment by LuigiHutch** [comment]
An (unclear?) scenario to where this is needed:

I am using the emerald pipe for its round robin feature as it will allow the items to be distributed evenly between auto workbenches so that the auto workbenchses dont have to wait for 1st stack of items to be pulled out to allow the second stack to be pulled out so it can start the crafting.

As the round robin functionis only found on an emerald pipe, this means that I have to use a filter to extract the items. This means that I cant change what is going to be crafted depending on what items are placed in the chests as you have to use all items that are in the filter.

Tl;DR... I am wanting to use the emerald pipes round robin feature without having to set a filter.

**Comment by SandGrainOne** [comment]
But when you craft stuff you need resources at a a very strict ratio. If you run out of one item the crafting will stop either way. This is the original purpose of the Emerald pipe and why it will stop and wait for items to become available again.

Say you want to make redstone repeaters. That will require 2 redstone torches, one redstone dust and 3 stone. In the emerald pipe you then use 2 filter slots for redstone torches, 1 filter slot for redstone dust and 3 for stone. The emerald pipe will then extract 2 torces, 1 dust and 3 stone and then repeat that over and over. It provides the perfect ratio of items and will stop and wait if something is missing.

**Comment by LuigiHutch** [comment]
Yeah but say if you want to craft a copper gear, you would just put the right amount of copper and stone gears in the chest for it to craft.

And say you want to craft a tin gear from the same chest, you would have to put the right amount of tin and stone gears in the chest for it to craft. 

As both these recipes require different materials, you would need 1 emerald pipe per recipe, (and say you could have quite a few recipes per chest)  which I find is quite unnecessary when you just need to use the pipes for its round robin feature.

**Comment by SandGrainOne** [comment]
Why don't you put the resources directly in the crafting machine or have the chest right next to the crafting machine (for those that support that). It sounds like the Emerald pipe is an unnecessary component in use cases like you describe. You have added a layer of automation you don't really need.

The perfect "machine" in situations like that would be something like the Forestry Worktable.

**Comment by AEnterprise** [comment]
the emerald pipe already has a blacklist mode, you can switch between blacklist and whitelist in the gui

**Comment by SandGrainOne** [comment]
@AEnterprise Both of those options extract all items from one slot before moving to the next slot. The feature being requested here would take one item of one type before taking one item of another valid type. 

The discussion we are having here now is to see if that distinction could be usefull.

**Comment by LuigiHutch** [comment]
You could use a worktable but if find that it is too manual and tedious. I want to put the items in the chest and go and do somthing else, then comeback when the crafting is complete.

Packager could be used (as it can pull items from adjacent chests) but it is kind of bulky as you have to route the items into a stamping table then back into the system.

(LONG)
The round robin has to be used as after the first item is crafted, the crafted item is routed depending on which auto workbench has items in for the next available craft. This means that if the item that has been crafted is not needed for any other craft, it is sent to the output chest. So if the first craft is already completed before all the items were fed into the the first auto workbench, the item would go into the output area instead of the second auto workbench as there is no way to detect if it needs to craft as there is no items in the auto work benches. The Items in inventory gate condition is used to tell if an item is needed to be crafted or sent to output

Here is a blueprint of my build of where it is necessary: https://goo.gl/ZekJ6p
(Needs BC + Forestry)

**Comment by asiekierka** [comment]
>Changing the 'mode' of an emerald pipe when the pipe is being powered

You mean "it only changes when you close the GUI". A very minor annoyance, but fixing it anyway...

**Comment by asiekierka** [comment]
The "bug" part fixed in https://github.com/BuildCraft/BuildCraft/commit/5ce9ab97f28fcd51280eafc1b103233d26a0eddc
